lib/dockerregistry/checks.go

```diff
@@ -245,3 +252,170 @@ func (check *ImageSizeCheck) Run() error {
+		sc *SyncContext,
+		reqs chan stream.ExternalRequest,
+		requestResults chan<- RequestResult,
+		wg *sync.WaitGroup,
+		mutex *sync.Mutex) {
+		for req := range reqs {
+			reqRes := RequestResult{Context: req}
+			errors := make(Errors, 0)
+			edge := req.RequestParams.(PromotionEdge)
+			occurrences, err := vulnProducer(edge)
+			if err != nil {
+				errors = append(errors, Error{
+					Context: "error getting vulnerabilities",
+					Error:   err})
+			}
+			if len(occurrences) > 0 {
+				for _, occ := range occurrences {
+					if CheckSeverity(occ, check.SeverityThreshold) {
+						errors = append(errors, Error{
+							Context: "vulnerabilities severity check",
+							Error: ImageVulnError{
+								edge.SrcImageTag.ImageName,
+								occ.GetType(),
+								int(occ.GetSeverity()),
+							},
+						})
+					}
+				}
+			}
+			reqRes.Errors = errors
+			requestResults <- reqRes
+		}
+	}
+
+	return check.SyncContext.ExecRequests(
+		populateRequests,
+		processRequest,
+	)
+}
+
+// Error is a function of ImageSizeError and implements the error interface.
+func (err ImageVulnError) Error() string {
+	return fmt.Sprintf("The image %q contains a vulnerability"+
+		"(type: %q, severity: %q)",
+		err.ImageName, err.VulnerabilityType, err.VulnerabilitySeverity)
+}
+
+// CheckSeverity checks if a vulnerability is a high enough severity to fail
+// the ImageVulnCheck.
+func CheckSeverity(
+	vuln *grafeaspb.VulnerabilityOccurrence,
+	severityThreshold int,
+) bool {
+	severityLevel := vuln.GetSeverity()
+	return int(severityLevel) >= severityThreshold
+}
+
+func parseImageProjectID(
+	edge PromotionEdge,
+) (string, error) {
+	const projectIDIndex = 1
+	splitName := strings.Split(string(edge.SrcRegistry.Name), "/")
+	if len(splitName) <= projectIDIndex {
+		return "", fmt.Errorf("Could not parse project ID from image name: %q",
+			string(edge.SrcRegistry.Name))
+	}
+	return splitName[projectIDIndex], nil
+}
+
+// mkRealVulnProducer returns an ImageVulnProducer that gets all vulnerability
+// Occurrences associated with the image represented in the PromotionEdge
+// using the Container Analysis Service client library.
+// nolint[errcheck]
+func mkRealVulnProducer(client *containeranalysis.Client) ImageVulnProducer {
+	return func(
+		edge PromotionEdge,
+	) ([]*grafeaspb.VulnerabilityOccurrence, error) {
+		// resourceURL is of the form https://gcr.io/[projectID]/my-image
+		resourceURL := "https://" + path.Join(string(edge.SrcRegistry.Name),
+			string(edge.SrcImageTag.ImageName))
+		projectID, err := parseImageProjectID(edge)
+		if err != nil {
+			return nil, fmt.Errorf("ParsingProjectID: %v", err)
+		}
+
+		ctx := context.Background()
+
+		req := &grafeaspb.ListOccurrencesRequest{
+			Parent: fmt.Sprintf("projects/%s", projectID),
+			Filter: fmt.Sprintf("resourceUrl = %q kind = %q",
+				resourceURL, "VULNERABILITY"),
+		}
+
+		var occurrenceList []*grafeaspb.VulnerabilityOccurrence
+		it := client.GetGrafeasClient().ListOccurrences(ctx, req)
```

Comment:
You should probably run GetGrafeasClient() on the outside, in block 303 somewhere. No need to call GetGrafeasClient() for every request, I think.

Reply:
GetGrafeasClient() just gets a field variable that already exists in the client (similar to calling client.grafeasClient), no extra logic. Is there any need to create another variable to hold the field variable?

Reply:
I was just thinking that you'd pass around not the client but the result of client.GetGrafeasClient(). No need to pass around 2 variables if you are only going to be using 1 of them.

But as we discussed offline, it's fine to keep this as-is, it's probably not a big deal. Up to you.